### ISSUE TITLE
modular bounds detection

### ISSUE THREAD

**Issue by NotJeven** [Moesh/Calamity#102]
z/x coord bounds are determined by some block detection at y=255, or y=128, something big to give a lot of vertical space to work with.
y coord (height) bounds are determined by some block detection at y=~2. I am using '~' to represent 'relative' as it works in game.
This will make huge structure zones but it gives level designers the ability to play with whatever maximum height they want.

This will keep level designers from having to hunt down script integration and make the scripts modular for each level. Each level will run on the same set of scripts.

In an example:
z/x will have a 1 block thick boundary layer over top the map at y=255. The game can constantly check if the player is outside the bounds by referencing the block at ~ 255 ~ from the player.

If we set the game to detect height bounds at ~ ~20 ~ from the player. The level designer can now use all the space from y=0 to y=253 (at y=253, ~ ~2 ~ will be at ~ 255 ~ where the x/z barriers are). If I make another layer of barrier blocks at y=100, the space from y=0 to y=98 will be usable to play in. 

To summarize. The level designer places 2 barrier layers over top their map. One at y=255, another 2 blocks above what they want the level height to be. The scripts are already checkout y=255 for x/z bounds, and y=~2 for height bounds. 

I am not sure if I explained this effectively or not, I have a hard time articulating ideas. I will write up a prototype once I understand the full scope of what the current map bound system is doing, unless there is something obvious I am totally missing here.

**Comment by Moesh** [comment]
Can you confirm if these are the appropriate actions for your request?

- Move x/z in-bound detection from yLevel 74 to yLevel 255
- Move x/y in-bound border detection from yLevel 73 to 254 [used as edge detection, needs its own layer]
- Move blue team x/y spawn zone from yLevel 71 to yLevel 253
- Move red team x/y spawn zone from yLevel 70 to yLevel 252
- New rule for accepting pull requests: Arena must have a method of erasing its entire play space

With these changes, levels should be able to have full control of the vertical space. @Vilder50 will have to speak to how void lane protection will be handled in this case. From a quick glace at the script, it seems to be thinking that level 67 (current max height) is the max height. This is because that is what we agreed on back when this feature was introduced.

This height limit extends out far past where the arena structures end.

Additional challenges for map makers:

- "In bound" barrier blocks will need to either be generated via commands or a separate array of structures will need to be created and controlled from the top. Modern arena uses straight-up rectangles to make it as easy as possible. This should be fine.
- Barrier markers are not visible from playable area, meaning there will be some abstraction required to make changes

I'm not particularly relishing doing this, but its a good idea, so why not?

**Comment by NotJeven** [comment]
Yes. Basically we make "data layers" way up in the sky that are loaded with structures for each map. We set these standards and say the usable space is lowest data layer - 10 or something. I am now imagining specific structure blocks *just* for the data layers.

I solved the height problem (above).

If we agree on the data layer standards, I can pull it all together.

**Comment by NotJeven** [comment]
This issue is addressed with #103. I closed #103 for reasons I addressed there.
